modules/serverpod_auth/serverpod_auth_server/lib/src/endpoints/sms_endpoint.dart

```diff
@@ -0,0 +1,129 @@
+
+import 'package:serverpod/serverpod.dart';
+import 'package:serverpod_auth_server/module.dart';
+import 'package:serverpod_auth_server/src/business/sms_auth.dart';
+
+/// Endpoint for handling Sign in with SMS.
+class SmsEndpoint extends Endpoint {
+  /// Start a Authentication with SMS. Returns a String containing a hash that
+  /// should be sent back to the server with the OTP. To verify.
+  Future<SmsOtpResponse> startAuthentication(
+    Session session,
+    String phoneNumber,
+  ) async {
+    if (await _hasTooManyFailedSignIns(session, phoneNumber)) {
+      session.log('Too many failed sign in attempts', level: LogLevel.warning);
+      return SmsOtpResponse(
+        success: false,
+        failReason: AuthenticationFailReason.tooManyFailedAttempts,
+      );
+    }
+    var hash = await SMSs.getOTPRequest(phoneNumber: phoneNumber);
+    return SmsOtpResponse(
+      success: true,
+      hash: hash,
+    );
+  }
+
+  /// Verifies that the OTP is correct and that the hash hasn't been tampered
+  /// with. If the OTP is correct, the user is logged in and a session is
+  /// returned.
+  Future<AuthenticationResponse> verifyAuthentication(
+    Session session,
+    String phoneNumber,
+    String otp,
+    String storedHash,
+  ) async {
+    session.log('authenticate $phoneNumber / XXXXXXXX', level: LogLevel.debug);
+
+    if (await _hasTooManyFailedSignIns(session, phoneNumber)) {
+      return AuthenticationResponse(
+          success: false,
+          failReason: AuthenticationFailReason.tooManyFailedAttempts);
+    }
+
+    var verification = await SMSs.verifyOTPRequest(
+      phoneNumber: phoneNumber,
+      otp: otp,
+      storedHash: storedHash,
+    );
+    if (!verification) {
+      session.log(' - invalid credentials', level: LogLevel.debug);
+      await _logFailedSignIn(session, phoneNumber);
+      return AuthenticationResponse(
+        success: false,
+        failReason: AuthenticationFailReason.invalidCredentials,
+      );
+    }
+
+    session.log(' - OTP verified', level: LogLevel.debug);
+
+    var entry = await session.db.findSingleRow<SmsAuth>(
+      where: SmsAuth.t.phoneNumber.equals(phoneNumber),
+    );
+
+    if (entry == null) {
+      session.log(' - no entry found', level: LogLevel.debug);
+      entry = await SMSs.createUser(session, phoneNumber);
+      if (entry == null) {
+        return AuthenticationResponse(
+          success: false,
+          failReason: AuthenticationFailReason.invalidCredentials,
+        );
+      }
+    }
+
+    session.log(' - user found', level: LogLevel.debug);
+
+    var userInfo = await Users.findUserByUserId(session, entry.userId);
+    // Check if user exists
+
+    // Create user if it doesn't exist and return it
+    if (userInfo == null) {
+      session.log('- user not fount', level: LogLevel.debug);
+      return AuthenticationResponse(
+        success: false,
+        failReason: AuthenticationFailReason.invalidCredentials,
+      );
+    }
```

Comment:
Question: The comment above doesn't seem to match what this code does.

Could we avoid multiple lookups for the user in the database? Maybe we could return it as a record from the `SMSs.createUser` method?